relay-kafka/src/producer/mod.rs

```diff
@@ -307,7 +325,24 @@ impl Producer {
         payload: &[u8],
     ) -> Result<&str, ClientError> {
         let now = Instant::now();
-        let topic_name = self.topic_name.as_str();
+
+        let (topic_name, producer, rate_limiter) = if self.topic_producers.len() == 1 {
+            &self.topic_producers[0]
+        } else {
+            let shard_index = match key {
+                Some(key_bytes) => {
+                    let mut hasher = FnvHasher::default();
+                    hasher.write(&key_bytes);
+                    hasher.finish32() as usize % self.topic_producers.len()
+                }
+                None => {
+                    // fetch_add wraps on overflow
+                    let count = self.round_robin_counter.fetch_add(1, Ordering::Relaxed);
+                    count % self.topic_producers.len()
```

Comment:
i think doing round-robin between shards in case of no key is probably cheaper than actually-random, wdyt